File: software/examples/perl/example_all_data.pl
#!/usr/bin/perl

use strict;
use Tinkerforge::IPConnection;
use Tinkerforge::BrickIMUV2;

use constant HOST => 'localhost';
use constant PORT => 4223;
use constant UID => 'XXYYZZ'; # Change XXYYZZ to the UID of your IMU Brick 2.0

# Callback subroutine for all data callback
sub cb_all_data
{
    my ($acceleration, $magnetic_field, $angular_velocity, $euler_angle, $quaternion,
        $linear_acceleration, $gravity_vector, $temperature, $calibration_status) = @_;

    print "Acceleration [X]: " . @{$acceleration}[0]/100.0 . " m/s²\n";
    print "Acceleration [Y]: " . @{$acceleration}[1]/100.0 . " m/s²\n";
    print "Acceleration [Z]: " . @{$acceleration}[2]/100.0 . " m/s²\n";
    print "Magnetic Field [X]: " . @{$magnetic_field}[0]/16.0 . " µT\n";
    print "Magnetic Field [Y]: " . @{$magnetic_field}[1]/16.0 . " µT\n";
    print "Magnetic Field [Z]: " . @{$magnetic_field}[2]/16.0 . " µT\n";
    print "Angular Velocity [X]: " . @{$angular_velocity}[0]/16.0 . " °/s\n";
    print "Angular Velocity [Y]: " . @{$angular_velocity}[1]/16.0 . " °/s\n";
    print "Angular Velocity [Z]: " . @{$angular_velocity}[2]/16.0 . " °/s\n";
    print "Euler Angle [X]: " . @{$euler_angle}[0]/16.0 . " °\n";
    print "Euler Angle [Y]: " . @{$euler_angle}[1]/16.0 . " °\n";
    print "Euler Angle [Z]: " . @{$euler_angle}[2]/16.0 . " °\n";
    print "Quaternion [W]: " . @{$quaternion}[0]/16383.0 . "\n";
    print "Quaternion [X]: " . @{$quaternion}[1]/16383.0 . "\n";
    print "Quaternion [Y]: " . @{$quaternion}[2]/16383.0 . "\n";
    print "Quaternion [Z]: " . @{$quaternion}[3]/16383.0 . "\n";
    print "Linear Acceleration [X]: " . @{$linear_acceleration}[0]/100.0 . " m/s²\n";
    print "Linear Acceleration [Y]: " . @{$linear_acceleration}[1]/100.0 . " m/s²\n";
    print "Linear Acceleration [Z]: " . @{$linear_acceleration}[2]/100.0 . " m/s²\n";
    print "Gravity Vector [X]: " . @{$gravity_vector}[0]/100.0 . " m/s²\n";
    print "Gravity Vector [Y]: " . @{$gravity_vector}[1]/100.0 . " m/s²\n";
    print "Gravity Vector [Z]: " . @{$gravity_vector}[2]/100.0 . " m/s²\n";
    print "Temperature: $temperature °C\n";
    print "Calibration Status: " . sprintf('%08b', $calibration_status) . "\n";
    print "\n";
}

my $ipcon = Tinkerforge::IPConnection->new(); # Create IP connection
my $imu = Tinkerforge::BrickIMUV2->new(&UID, $ipcon); # Create device object

$ipcon->connect(&HOST, &PORT); # Connect to brickd
# Don't use device before ipcon is connected

# Register all data callback to subroutine cb_all_data
$imu->register_callback($imu->CALLBACK_ALL_DATA, 'cb_all_data');

# Set period for all data callback to 0.1s (100ms)
$imu->set_all_data_period(100);

print "Press key to exit\n";
<STDIN>;
$ipcon->disconnect();

File: software/examples/javascript/ExampleAllData.js
var Tinkerforge = require('tinkerforge');

var HOST = 'localhost';
var PORT = 4223;
var UID = 'XXYYZZ'; // Change XXYYZZ to the UID of your IMU Brick 2.0

var ipcon = new Tinkerforge.IPConnection(); // Create IP connection
var imu = new Tinkerforge.BrickIMUV2(UID, ipcon); // Create device object

ipcon.connect(HOST, PORT,
    function (error) {
        console.log('Error: ' + error);
    }
); // Connect to brickd
// Don't use device before ipcon is connected

ipcon.on(Tinkerforge.IPConnection.CALLBACK_CONNECTED,
    function (connectReason) {
        // Set period for all data callback to 0.1s (100ms)
        imu.setAllDataPeriod(100);
    }
);

// Register all data callback
imu.on(Tinkerforge.BrickIMUV2.CALLBACK_ALL_DATA,
    // Callback function for all data callback
    function (acceleration, magneticField, angularVelocity, eulerAngle, quaternion,
              linearAcceleration, gravityVector, temperature, calibrationStatus) {
        console.log('Acceleration [X]: ' + acceleration[0]/100.0 + ' m/s²');
        console.log('Acceleration [Y]: ' + acceleration[1]/100.0 + ' m/s²');
        console.log('Acceleration [Z]: ' + acceleration[2]/100.0 + ' m/s²');
        console.log('Magnetic Field [X]: ' + magneticField[0]/16.0 + ' µT');
        console.log('Magnetic Field [Y]: ' + magneticField[1]/16.0 + ' µT');
        console.log('Magnetic Field [Z]: ' + magneticField[2]/16.0 + ' µT');
        console.log('Angular Velocity [X]: ' + angularVelocity[0]/16.0 + ' °/s');
        console.log('Angular Velocity [Y]: ' + angularVelocity[1]/16.0 + ' °/s');
        console.log('Angular Velocity [Z]: ' + angularVelocity[2]/16.0 + ' °/s');
        console.log('Euler Angle [X]: ' + eulerAngle[0]/16.0 + ' °');
        console.log('Euler Angle [Y]: ' + eulerAngle[1]/16.0 + ' °');
        console.log('Euler Angle [Z]: ' + eulerAngle[2]/16.0 + ' °');
        console.log('Quaternion [W]: ' + quaternion[0]/16383.0);
        console.log('Quaternion [X]: ' + quaternion[1]/16383.0);
        console.log('Quaternion [Y]: ' + quaternion[2]/16383.0);
        console.log('Quaternion [Z]: ' + quaternion[3]/16383.0);
        console.log('Linear Acceleration [X]: ' + linearAcceleration[0]/100.0 + ' m/s²');
        console.log('Linear Acceleration [Y]: ' + linearAcceleration[1]/100.0 + ' m/s²');
        console.log('Linear Acceleration [Z]: ' + linearAcceleration[2]/100.0 + ' m/s²');
        console.log('Gravity Vector [X]: ' + gravityVector[0]/100.0 + ' m/s²');
        console.log('Gravity Vector [Y]: ' + gravityVector[1]/100.0 + ' m/s²');
        console.log('Gravity Vector [Z]: ' + gravityVector[2]/100.0 + ' m/s²');
        console.log('Temperature: ' + temperature + ' °C');
        console.log('Calibration Status: ' + calibrationStatus.toString(2));
        console.log();
    }
);

console.log('Press key to exit');
process.stdin.on('data',
    function (data) {
        ipcon.disconnect();
        process.exit(0);
    }
);
